File: study/lua/src/app/scene/view/rollnotice/RollNoticeBaseNode.lua

local RollNoticeHelper = require("app.scene.view.rollnotice.RollNoticeHelper")

local RollNoticeBaseNode = class("RollNoticeBaseNode", function ()
	return display.newNode()
end)

function RollNoticeBaseNode:ctor(rollMsg)
	self._rollMsg = rollMsg
	self:setCascadeOpacityEnabled(true)
	-- 富文本
    local richTxt = RollNoticeHelper.makeRichMsgFromServerRollMsg(rollMsg)
	local richText = ccui.RichText:createWithContent(richTxt)
	richText:setCascadeOpacityEnabled(true)
	richText:formatText()
	local richTextSize = richText:getVirtualRendererSize()
	self:addChild(richText)
	self:setContentSize(richTextSize)
	richText:setPosition(richTextSize.width * 0.5,richTextSize.height * 0.5)
end

function RollNoticeBaseNode:run(containSize)
	local conSize = self:getContentSize()
	local y = containSize.height * 0.5
	local targetPos = cc.p(0-conSize.width/2-50,y)
	local fromPos = cc.p(containSize.width + conSize.width/2 + 50,y)
	self:setAnchorPoint(cc.p(0.5,0.5))
	self:setPosition(fromPos.x, fromPos.y)
	local duration = math.abs(targetPos.x - fromPos.x)/640*4
	local moveAction = cc.MoveTo:create(duration,targetPos)

	self:setVisible(false)
	self:runAction(cc.Sequence:create(
		-- 随机延迟
		cc.Show:create(),
		moveAction,
		cc.CallFunc:create(function(node)
			print("remove from root_node -------------------------------------")
			G_SignalManager:dispatch(SignalConst.EVENT_SUBTITLES_RUN_END,self._rollMsg,node)
		end),
		cc.RemoveSelf:create()
	))
end

return RollNoticeBaseNode

File: study/ts/script/scene/view/rollnotice/RollNoticeBaseNode.ts
import RollNoticeHelper from "./RollNoticeHelper";
import { RichTextExtend } from "../../../extends/RichTextExtend";
import { G_SignalManager } from "../../../init";
import { SignalConst } from "../../../const/SignalConst";
import { handler } from "../../../utils/handler";

export default class RollNoticeBaseNode extends cc.Component {
    _rollMsg: any;
    // _node: cc.Node;
    initData(rollMsg) {
        this._rollMsg = rollMsg;
        // this.setCascadeOpacityEnabled(true);
        var richTxt = RollNoticeHelper.makeRichMsgFromServerRollMsg(rollMsg, null);
        var richText = RichTextExtend.createWithContent(richTxt[0]);
        // richText.setCascadeOpacityEnabled(true);
        // richText.formatText();
        this.node.addChild(richText.node);
        var richTextSize = richText.node.getContentSize();
        this.node.setContentSize(richTextSize);
        richText.node.setPosition(richTextSize.width * 0.5,4);
    }
    run(containSize) {
        var conSize = this.node.getContentSize();
        var y = 0;
        var targetPos = cc.v2(0 - conSize.width - 50 - containSize.width * 0.5, y);
        var fromPos = cc.v2(containSize.width * 0.5 + 50, y);
        this.node.setAnchorPoint(cc.v2(0.5, 0.5));
        this.node.setPosition(fromPos.x, fromPos.y);
        var duration = Math.abs(targetPos.x - fromPos.x) / 640 * 4;
        var moveAction = cc.moveTo(duration, targetPos);
        this.node.active = (true);

        var callFunc = cc.callFunc(handler(this, this.onFinish));
        this.node.runAction(cc.sequence(moveAction, callFunc));
    }

    onFinish(node) {
        G_SignalManager.dispatch(SignalConst.EVENT_SUBTITLES_RUN_END, this._rollMsg, node);
        this.node.removeFromParent();
    }
}